### DISCUSSION TITLE
`Sync` in handler, future cannot be sent between threads safely

### DISCUSSION THREAD

**Discussion by allan2** [tokio-rs/axum#1708] [Q&A] [answered]
I'm trying to dynamically construct Postgres query parameters. The Postgres params type is [`&[&(dyn ToSql + Sync)]`](https://docs.rs/postgres/latest/postgres/struct.Client.html#method.execute).

I tried this:
```rust
async fn insert_multirow_once(DbConn(conn): DbConn) -> Result<impl IntoResponse, StatusCode> {
    let statement = "INSERT INTO foo (a, b) VALUES ($1, $2), ($3, $4)";

    let mut params = Vec::<Box<dyn ToSql + Sync>>::with_capacity(4);

    for i in 1..=2 {
        params.push(Box::new(i));
        params.push(Box::new("s".to_owned()));
    }

    // from https://github.com/sfackler/rust-postgres/issues/712
    let params = params.iter().map(|x| x.as_ref()).collect::<Vec<_>>();

    conn.execute(statement, &params)
        .await
        .map_err(internal_error)?;
    Ok(StatusCode::OK)
}
```

The function by itself is ok but it doesn't work as a handler. The error is `future cannot be sent between threads safely`.

I searched the issues for `Send` and `Sync` but I'm still not sure what to do. Any help would be appreciated.

[Full example here](https://github.com/allan2/axum-postgres-dynamic-params), based on the tokio-postgres example

**Comment by davidpdrsn** [comment]
Hm I'm not sure. Try asking the tokio-postgres maintainers.

**Comment by allan2** [comment, marked as answer]
I got it working. I had read the [solution](https://github.com/sfackler/rust-postgres/issues/712#issuecomment-743456104) but forgot to actually cast.

Changed this:
```rust
let mut params = Vec::<Box<dyn ToSql + Sync + Send>>::with_capacity(4);  // add Send
```

and this
```rust
let params = params
    .iter()
    .map(|x| x.as_ref() as &(dyn ToSql + Sync))  // need to cast
    .collect::<Vec<_>>();
 ```
 and now it works.

Thanks for Axum!
